### DISCUSSION TITLE
Working with MosaicJSON and multiple bands in different files

### DISCUSSION THREAD

**Discussion by Duke74** [developmentseed/titiler#1037] [Q&A]
Hi,

Thanks for this awesome project, it's really helpfull.

I've been doing some tests with sentinel2 data lately and the tiler is working great with a mosaicJSON file that link directly to cog files like this : 

> {
    "mosaicjson": "0.0.3",
    "version": "1.0.0",
    "minzoom": 11,
    "maxzoom": 15,
    "quadkey_zoom": 11,
    "bounds": [-127.87505894081966, 23.562648562762604, -111.8838502136324, 41.55183566226867],
    "center": [-119.87945457722603, 32.55724211251564, 11],
    "tiles": {
        "02122323032": ["https://sentinel-cogs.s3.us-west-2.amazonaws.com/sentinel-s2-l2a-cogs/9/T/XF/2024/6/S2A_9TXF_20240628_0_L2A/TCI.tif"],
        "02122323033": ["https://sentinel-cogs.s3.us-west-2.amazonaws.com/sentinel-s2-l2a-cogs/9/T/XF/2024/6/S2A_9TXF_20240628_0_L2A/TCI.tif"],
...,

Now rather than using the TCI sentinel files I would like to work with the different cogs available for each items exemple here: 
https://radiantearth.github.io/stac-browser/#/external/earth-search.aws.element84.com/v1/collections/sentinel-2-l2a/items/S2B_32TLR_20241104_0_L2A

Like for exemple merging the red, green ,blue from the different single band cogs available, rather than using the TCI file

But my issue is I'm not sure how to do it. The mosaicJSON spec doesn't seems to make it possible to have multiple files for multiple bands for a same tile.
Can mosaicJSON link directly to the STAC items ? 
But then what parameters to use when requesting the tiles ?

Thanks for the help !

**Comment by vincentsarago** [comment]
Hi @Duke74 

> Can mosaicJSON link directly to the STAC items ?

Yes, but this takes customization because you'll need a `dataset_reader` https://github.com/developmentseed/titiler/blob/72a8df9510895405e183ecb2b41df45ce1d4f089/src/titiler/mosaic/titiler/mosaic/factory.py#L83-L87

to be of type `MultiBandReader` and some other dependencies customization. 

something like 👇 
```python
from typing import Type
import attr

from fastapi import FastAPI

from rio_tiler.io import MultiBandReader, BaseReader, Reader
from rio_tiler.errors import InvalidBandName

from titiler.mosaic.factory import MosaicTilerFactory

app = FastAPI(openapi_url="/api", docs_url="/api.html",)

@attr.s
class customReader(MultiBandReader):

    reader: Type[BaseReader] = attr.ib(default=Reader, init=False)
    
    def __attrs_post_init__(self):
        """Fetch Reference band to get the bounds."""
        self.bands = [f"b{ix + 1}" for ix in range(len(self.input))]

        # We set the bands to be loaded by default to be all the bands
        self.default_bands = [f"b{ix + 1}" for ix in range(len(self.input))]

        with self.reader(self.input[0], tms=self.tms, **self.reader_options) as cog:
            self.bounds = cog.bounds
            self.crs = cog.crs
            self.minzoom = cog.minzoom
            self.maxzoom = cog.maxzoom

    def _get_band_url(self, band: str) -> str:
        """Validate band's name and return band's url."""
        if band not in self.bands:
            raise InvalidBandName(f"{band} is not valid")

        index = self.bands.index(band)
        return self.input[index]


mosaic = MosaicTilerFactory(dataset_reader=customReader, router_prefix="/mosaicjson")
app.include_router(
    mosaic.router,
    prefix="/mosaicjson",
    tags=["MosaicJSON"],
)
```
